src/nfa/nfa.cc

```diff
@@ -788,110 +1071,21 @@ void Mata::Nfa::uni(Nfa *unionAutomaton, const Nfa &lhs, const Nfa &rhs) {
 }
 
 void Mata::Nfa::intersection(Nfa *res, const Nfa &lhs, const Nfa &rhs, ProductMap*  prod_map) {
-    using StatePair = std::pair<State, State>;
-
-    /* If q is the state of this automaton and p of other, then thisAndOtherStateToIntersectState[q][p] = state in intersect
-     * representing (q,p). The hash function computes for each pair (q,p) unique number (q + p*(num of states in this automaton))
-     * whose std hash is returned.
-     *
-     * see https://stackoverflow.com/questions/15719084/how-to-use-lambda-function-as-hash-function-in-unordered-map
-     */
-    auto hashStatePair = [lhs](const StatePair &sp) {
-        return std::hash<unsigned long>()(sp.first + sp.second*lhs.transitionrelation.size());
-    };
-    // TODO probably remove this since we use prod_map as parameter
-    //std::unordered_map<StatePair, State, decltype(hashStatePair)> thisAndOtherStateToIntersectState(10, hashStatePair); // TODO default buckets?
-
-    if (prod_map == nullptr) {
-        prod_map = new ProductMap();
-    }
-
-    std::vector<StatePair> pairsToProcess;
-
-    for (State thisInitialState : lhs.initialstates) {
-        for (State otherInitialState : rhs.initialstates) {
-            StatePair thisAndOtherInitialStatePair(thisInitialState, otherInitialState);
-            State newIntersectState = res->add_new_state();
-
-            (*prod_map)[thisAndOtherInitialStatePair] = newIntersectState;
-            pairsToProcess.push_back(thisAndOtherInitialStatePair);
-
-            res->make_initial(newIntersectState);
-            if (lhs.has_final(thisInitialState) && rhs.has_final(otherInitialState))
-                res->make_initial(newIntersectState);
-            if (lhs.has_final(thisInitialState) && rhs.has_final(otherInitialState)) {
-                res->make_final(newIntersectState);
-            }
-        }
+    auto intersection { Intersection(lhs, rhs) };
```

Comment:
Seeing the usage, I'm wondering, whether this could maybe be encapsulated in some static method instead? Like `Interesection.compute(lhs, rhs)`. @martinhruska  what do you think?

Reply:
I have considered this approach, too. They achieve the same and I am OK with both approaches. I will go with whatever you prefer.

Reply:
Or maybe, the ultimate snippet would be: `Intersection.of(lhs, rhs)`. I'll leave the decision to Martin and others, though.

Reply:
Could be, too. I think that `.compute()` is a more standard approach to this than `.of()`, nevertheless, both are perfectly usable in my opinion.

Reply:
Intersection.compute() would be fine.

Reply:
Sounds good to me, too. `Intersction.compute()` it is, then.